packages/studio-plugin/src/HmrManager.ts

```diff
@@ -1,78 +1,65 @@
 import ParsingOrchestrator from "./ParsingOrchestrator";
 import { StudioHMRPayload, StudioHMRUpdateID, UserPaths } from "./types";
-import { HmrContext, ViteDevServer } from "vite";
+import { ViteDevServer } from "vite";
 import VirtualModuleID from "./VirtualModuleID";
+import chokidar from "chokidar";
 import upath from "upath";
 
 /**
  * HmrManager is responsible for handling studio specific HMR updates.
  */
 export default class HmrManager {
   constructor(
+    private server: ViteDevServer,
     private orchestrator: ParsingOrchestrator,
-    private pathToUserProjectRoot: string,
     private userPaths: UserPaths
   ) {}
 
+  createWatcher(pathToUserProjectRoot: string) {
+    const watcher = chokidar.watch(upath.join(pathToUserProjectRoot, "src"));
+    watcher.on("change", this.handleHotUpdate);
+    watcher.on("unlink", this.handleHotUpdate);
+  }
+
   /**
    * A custom handler for vite hot updates.
    *
    * See import('vite').Plugin.handleHotUpdate
    */
-  handleHotUpdate = async (ctx: HmrContext): Promise<void> => {
-    const { server, file } = ctx;
-
-    await HmrManager.reloadAssociatedModules(ctx);
-    if (!file.startsWith(this.pathToUserProjectRoot)) {
-      return;
-    }
+  handleHotUpdate = (unnormalizedFilepath: string) => {
+    const file = upath.normalize(unnormalizedFilepath);
     this.orchestrator.reloadFile(file);
-    HmrManager.invalidateStudioData(server);
+    this.invalidateStudioData();
     const updateType = getHMRUpdateType(file, this.userPaths);
-    const data = this.getPayload(updateType);
-    server.ws.send({
+    const data = this.getPayload(updateType, file);
+    this.server.ws.send({
       type: "custom",
       event: StudioHMRUpdateID,
       data,
     });
   };
 
-  /**
-   * Whether or not a given file should be excluded from being watched.
-   *
-   * Currently we ignore all files under localData, to avoid a full page refresh when
-   * the generate-test-data yext CLI command is called
-   */
-  shouldExcludeFromWatch = (filepath: string): boolean => {
-    return upath.normalize(filepath).startsWith(this.userPaths.localData);
-  };
-
-  private static async reloadAssociatedModules(ctx: HmrContext) {
-    const reloadModulePromises = ctx.modules.map((m) => {
-      return ctx.server.reloadModule(m);
```

Comment:
what was this needed for before?

Reply:
This was emulating default hmr behavior
I think it might have been possible before to just return undefined if we wanted default behavior

Reply:
would we still want default hmr behavior for files in the project root that Studio doesn't care about?

Reply:
I'm confused, I thought `reloadAssociatedModules` was the default HMR behavior that we had for all files (except `localData` files) and we just had extra behavior if the file was in the project root. Now, HMR for files outside the project root is handled by Vite and HMR for files inside the `src` directory in the project root is handled by `handleHotUpdate`. But I think files in the project root, but not in the `src` directory, wouldn't have any HMR handling now, when we used to reload the modules before?

Reply:
it would be called before but wouldnt functionally do anything because no other modules would be registered as importers

Reply:
We also want to ignore localData and just generally speaking vite's HMR will not work properly with any of the user's own files

Reply:
I might be missing something but I'm not sure how that makes it more clear there's only two cases?

Reply:
ah I forgot about localData. makes sense as is then!